src/engine/ResultTable.h

```diff
@@ -27,59 +28,85 @@ using std::mutex;
+// `ResultTable` and an `IdTable` is.
 class ResultTable {
  public:
+  // The status of the result.
   enum Status { IN_PROGRESS = 0, FINISHED = 1, ABORTED = 2 };
-  using ResultType = qlever::ResultType;
 
-  /**
-   * @brief This vector contains a list of column indices by which the result
-   *        is sorted. This vector may be empty if the result is not sorted
-   *        on any column. The primary sort column is _sortedBy[0]
-   *        (if it exists).
-   */
-  vector<size_t> _sortedBy;
+ public:
+  // TODO: I think that none of these member variables should be public. They
+  // probably are for historical reasons. I already added a `getLocalVocab`
+  // method.
 
+  // The actual entries.
   IdTable _idTable;
 
+  // The result type for each column. TODO: What if a column contains results of
+  // various types?
+  using ResultType = qlever::ResultType;
   vector<ResultType> _resultTypes;
 
-  std::shared_ptr<LocalVocab> _localVocab;
-
-  explicit ResultTable(ad_utility::AllocatorWithLimit<Id> allocator);
+  // The column indices by which the result is sorted (primary sort key first).
+  // Empty if the result is not sorted on any column.
+  vector<size_t> _sortedBy;
 
-  ResultTable(const ResultTable& other) = delete;
+ private:
+  // The local vocabulary of the result.
+  std::shared_ptr<LocalVocab> localVocab_ = std::make_shared<LocalVocab>();
 
-  ResultTable(ResultTable&& other) = default;
+ public:
+  // Construct with given allocator.
+  explicit ResultTable(ad_utility::AllocatorWithLimit<Id> allocator)
+      : _idTable(std::move(allocator)) {}
 
+  // Prevent accidental copying of a result table.
+  ResultTable(const ResultTable& other) = delete;
   ResultTable& operator=(const ResultTable& other) = delete;
 
+  // Moving of a result table is OK.
+  ResultTable(ResultTable&& other) = default;
   ResultTable& operator=(ResultTable&& other) = default;
 
-  virtual ~ResultTable();
+  // Default destructor.
+  virtual ~ResultTable() = default;
+
+  // Get the number of rows of this result.
+  size_t size() const { return _idTable.size(); }
 
-  size_t size() const;
+  // Get the number of columns of this result.
   size_t width() const { return _idTable.numColumns(); }
 
-  // Log to INFO the size of this result.
-  //
-  // NOTE: Due to the current sub-optimal design of `Server::processQuery`, we
-  // need the same message in multiple places and so instead of duplicating the
-  // message, we should have a method for it.
+  // Set local vocabulary. This only copies a shared pointer. In order to make a
+  // deep copy, use LocalVocab::clone(); see `GroupBy.cpp` for an example.
+  void setLocalVocab(const std::shared_ptr<LocalVocab>& localVocab) {
+    localVocab_ = localVocab;
+  }
+
+  // Get the local vocabulary of this result (not `const` because we may want to
+  // augment it while constructing the result).
+  std::shared_ptr<LocalVocab> getLocalVocab() const { return localVocab_; }
```

Comment:
I also thought about other interfaces. But isn't one reason for the shared pointer exactly that it can be shared by multiple results and that we want to clean it up when it's no longer used by any result?

Reply:
Thanks for the opportunity to clarify my intent:
I absolutely agree with the internal management of the local vocab via a `shared_ptr`.
I just wanted to discuss the way that the interface of the `ResultTable` class exposes this sharedness or not.
For the "normal" handling (reading and writing words) we don't need to expose it, so it can just be
`LocalVocab& localVocab() { return *sharedPtrToLocalVocab_;}`.

The places where we need the shared semantics (which are a bit more dangerous, because you can make more mistakes) are limited, so we can make the methods more explicit:

```
void ResultTable::shareLocalVocabFrom(const ResultTable& other) {
  sharedPtrToLocalVocab = other.sharedPtrToLocalVocab;
}
```


The only thing I don't completely like here is again the shallow constness
(You can logically modify `other` by sharing its local vocab and then modifying via the `shared_ptr`. But this is an underlying issue of our implementation which we have in ALL the variants (existing as well as proposed). If you really want this then we can implement an even safer interface that enforces "I am not allowed to modify my local vocab if it is share with some other operation", but I am not insisting on it.

Although... With more thinking, it can be done by simply setting a bool and adding an `AD_CONTRACT_CHECK` into the function that adds words (the bool then has to live within the local vocab). This function deals with strings etc. anyway, so we can afford the CHECK.